### PR TITLE
StackPromotion: support promoting allocations in dead-end control flow regions

### PR DESCRIPTION
**Pull request by eeckstein**
rdar://109274869

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci benchmark

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci test

File: SwiftCompilerSources/Sources/Optimizer/FunctionPasses/StackPromotion.swift [hunk 1]
```diff
@@ -75,6 +67,41 @@ private func tryPromoteAlloc(_ allocRef: AllocRefInstBase,
...
+    // missing, because otherwise the object would be leaking. For example:
+    //
+    //  bb1:
+    //    %k = alloc_ref $Klass
+    //    ...                    // %k's lifetime must end somewhere here
+    //    cond_br %c, bb1, bb2
+    //  bb2:
+    //    unreachable
+    //
+    // Therefore, it the allocation is inside a loop, we can treat it like allocations in
```
Comment (LucianoPAlmeida) [inline SwiftCompilerSources/Sources/Optimizer/FunctionPasses/StackPromotion.swift:94]: Typo `it the allocation` -> `if the allocation`?

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Pull request merged**